File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:

    def threeSum(self, nums):

        n = len(nums)

        # Step 1 : Sort the array
        nums.sort()

        # Final answer list
        ans = []

        # Step 2 : Fix one element
        for i in range(n):

            # Skip duplicate elements
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left = i + 1
            right = n - 1

            # Two Pointer Approach
            while left < right:

                total = nums[i] + nums[left] + nums[right]

                # If sum is smaller
                if total < 0:
                    left += 1

                # If sum is larger
                elif total > 0:
                    right -= 1

                # Valid triplet found
                else:

                    ans.append([
                        nums[i],
                        nums[left],
                        nums[right]
                    ])

                    left += 1
                    right -= 1

                    # Skip duplicate values
                    while (left < right and
                           nums[left] == nums[left - 1]):
                        left += 1

                    while (left < right and
                           nums[right] == nums[right + 1]):
                        right -= 1

        return ans

File: Data Structures & Algorithms/majority-element/submission-2.java
class Solution {

    public int majorityElement(int[] nums) {

        int fq = 0;
        int ans = 0;

        for (int i = 0; i < nums.length; i++) {

            if (fq == 0) {
                ans = nums[i];
            }

            if (ans == nums[i]) {
                fq++;
            } else {
                fq--;
            }
        }

        int count = 0;

        for (int val : nums) {
            if (val == ans) {
                count++;
            }
        }

        return count > nums.length / 2 ? ans : -1;
    }
}